packages/shared_preferences/shared_preferences_platform_interface/lib/in_memory_shared_preferences_async.dart

```diff
@@ -0,0 +1,158 @@
+  InMemorySharedPreferencesAsync.empty() : _data = <String, Object>{};
+
+  /// Instantiates an in-memory preferences store containing a copy of [data].
+  InMemorySharedPreferencesAsync.withData(Map<String, Object> data)
+      : _data = Map<String, Object>.from(data);
+
+  final Map<String, Object> _data;
+
+  @override
+  Future<bool> clear(
+    ClearPreferencesParameters parameters,
+    SharedPreferencesOptions options,
+  ) async {
+    final PreferencesFilters filter = parameters.filter;
+    if (filter.allowList != null) {
+      _data.removeWhere((String key, _) => filter.allowList!.contains(key));
+    } else {
+      _data.clear();
+    }
+    return true;
+  }
+
+  @override
+  Future<Map<String, Object>> getPreferences(
+    GetPreferencesParameters parameters,
+    SharedPreferencesOptions options,
+  ) async {
+    final PreferencesFilters filter = parameters.filter;
+    final Map<String, Object> preferences = Map<String, Object>.from(_data);
+    preferences.removeWhere((String key, _) =>
+        filter.allowList != null && !filter.allowList!.contains(key));
+    return preferences;
+  }
+
+  Future<bool> _setValue(
+    String key,
+    Object value,
+    SharedPreferencesOptions options,
+  ) async {
+    _data[key] = value;
+    return true;
+  }
+
+  @override
+  Future<bool> setString(
+    String key,
+    String value,
+    SharedPreferencesOptions options,
+  ) async {
+    return _setValue(key, value, options);
+  }
+
+  @override
+  Future<bool> setInt(
+    String key,
+    int value,
+    SharedPreferencesOptions options,
+  ) async {
+    return _setValue(key, value, options);
+  }
+
+  @override
+  Future<bool> setDouble(
+    String key,
+    double value,
+    SharedPreferencesOptions options,
+  ) async {
+    return _setValue(key, value, options);
+  }
+
+  @override
+  Future<bool> setBool(
+    String key,
+    bool value,
+    SharedPreferencesOptions options,
+  ) async {
+    return _setValue(key, value, options);
+  }
+
+  @override
+  Future<bool> setStringList(
+    String key,
+    List<String> value,
+    SharedPreferencesOptions options,
+  ) async {
+    return _setValue(key, value, options);
+  }
+
+  @override
+  Future<String?> getString(
+    String key,
+    SharedPreferencesOptions options,
+  ) async {
+    return _data[key] as String?;
+  }
+
+  @override
+  Future<bool?> getBool(
+    String key,
+    SharedPreferencesOptions options,
+  ) async {
+    return _data[key] as bool?;
+  }
+
+  @override
+  Future<double?> getDouble(
+    String key,
+    SharedPreferencesOptions options,
+  ) async {
+    return _data[key] as double?;
+  }
+
+  @override
+  Future<int?> getInt(
+    String key,
+    SharedPreferencesOptions options,
+  ) async {
+    return _data[key] as int?;
+  }
+
+  @override
+  Future<List<String>?> getStringList(
+    String key,
+    SharedPreferencesOptions options,
+  ) async {
+    final List<Object?>? data = _data[key] as List<Object?>?;
```

Comment:
`final List<Object>? data = _data[key] as List<Object>?;`

Whatever type the list has should be non-nullable, so I believe that will work at runtime. (In fact, for this particular implementation I think just `return _data[key] as List<String>?` would be fine since this is staying within Dart, so the value should never lose the original `List<String>` runtime type.)